src/lm/participants.ts

```diff
@@ -0,0 +1,196 @@
+				} else if (part instanceof vscode.LanguageModelToolCallPart) {
+
+					const tool = vscode.lm.tools.find(tool => tool.name === part.name);
+					if (!tool) {
+						throw new Error('Got invalid tool choice: ' + part.name);
+					}
+
+					let parameters: any;
+					try {
+						parameters = part.parameters;
+					} catch (err) {
+						throw new Error(`Got invalid tool use parameters: "${JSON.stringify(part.parameters)}". (${(err as Error).message})`);
+					}
+
+					const invocationOptions = { parameters, toolInvocationToken: request.toolInvocationToken, requestedContentTypes: ['text/plain', 'text/markdown', 'text/json', 'text/display'] };
+					toolCalls.push({
+						call: part,
+						result: vscode.lm.invokeTool(tool.name, invocationOptions, token),
+						tool
+					});
+				}
+			}
+
+			if (toolCalls.length) {
+				const assistantMsg = vscode.LanguageModelChatMessage.Assistant('');
+				assistantMsg.content2 = toolCalls.map(toolCall => new vscode.LanguageModelToolCallPart(toolCall.tool.name, toolCall.call.toolCallId, toolCall.call.parameters));
+				this.state.addMessage(assistantMsg);
+
+				let hasJson = false;
+				let display: string | undefined;
+				for (const toolCall of toolCalls) {
+					let toolCallResult = (await toolCall.result);
+
+					const plainText = toolCallResult['text/plain'];
+					const markdown = toolCallResult['text/markdown'];
+					const json = toolCallResult['text/json'];
+					display = toolCallResult['text/display']; // our own fake type that we use to indicate something that should be streamed to the user
+					if (display) {
+						stream.markdown(display);
+					}
+
+					const content: (string | vscode.LanguageModelToolResultPart | vscode.LanguageModelToolCallPart)[] = [];
+					let isOnlyPlaintext = true;
+					if (json !== undefined) {
+						content.push(new vscode.LanguageModelToolResultPart(toolCall.call.toolCallId, JSON.stringify(json)));
+						isOnlyPlaintext = false;
+						hasJson = true;
+
+					} else if (markdown !== undefined) {
+						content.push(new vscode.LanguageModelToolResultPart(toolCall.call.toolCallId, markdown));
+						isOnlyPlaintext = false;
+					}
+					if (plainText !== undefined) {
+						if (isOnlyPlaintext) {
+							content.push(new vscode.LanguageModelToolResultPart(toolCall.call.toolCallId, plainText));
+						} else {
+							content.push(plainText);
+						}
+					}
+					const message = vscode.LanguageModelChatMessage.User('');
+					message.content2 = content;
+					this.state.addMessage(message);
+				}
+
+				this.state.addMessage(vscode.LanguageModelChatMessage.User(`Above is the result of calling the functions ${toolCalls.map(call => call.tool.name).join(', ')}.${hasJson ? ' The JSON is also included and should be passed to the next tool.' : ''} ${display ? 'The user can see the result of the tool call and doesn\'t need you to show it.' : 'The user cannot see the result of the tool call, so you should show it to them in an appropriate way.'}`));
```

Comment:
Localize?

Reply:
Leaving in English since this goes to the llm, not to the user.